### PR TITLE
ethstats: stop loop if subscription fail

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**

### CODE REVIEW

File: ethstats/ethstats.go [hunk 1]
```diff
@@ -185,12 +185,17 @@ func (s *Service) loop() {
...
 	txSub := s.backend.SubscribeNewTxsEvent(txEventCh)
-	defer txSub.Unsubscribe()
 
+	if headSub == nil || txSub == nil {
+		log.Info("Stats daemon stopped due to nil subscription")
+		return
+	}
+
+	defer headSub.Unsubscribe()
+	defer txSub.Unsubscribe()
```
Comment (holiman) [inline ethstats/ethstats.go:198]: I think you should do it one at a time. If headsub is non-nil, you still should unsubscribe from it

**Review by alisscco — APPROVED** [review]

File: ethstats/ethstats.go [hunk 2]
```diff
@@ -185,10 +185,18 @@ func (s *Service) loop() {
 	// Subscribe to chain events to execute updates on
 	chainHeadCh := make(chan core.ChainHeadEvent, chainHeadChanSize)
 	headSub := s.backend.SubscribeChainHeadEvent(chainHeadCh)
+	if headSub == nil {
+		log.Info("Stats daemon stopped due to nil head subscription")
+		return
+	}
```
Comment (fjl) [inline ethstats/ethstats.go:191]: It's easy to avoid this error subscribing earlier. If you create the subscriptions in `New`, they can never be nil.

**Review by alisscco — APPROVED** [review]

File: ethstats/ethstats.go (hunk 2 above)
Comment (holiman) [reply ethstats/ethstats.go:191]: If they're created in `Start`, then the deferred `Unsubscribe` will become "decoupled" from the Subscribe, which is not great either...

**Review by holiman — APPROVED** [review]
LGTM

File: ethstats/ethstats.go (hunk 2 above)
Comment (fjl) [reply ethstats/ethstats.go:191]: The Unsubscribe would need to move into Service.Stop then.

**Review by alisscco — APPROVED** [review]

File: ethstats/ethstats.go [hunk 3]
```diff
@@ -167,38 +169,30 @@ func New(node *node.Node, backend backend, engine consensus.Engine, url string)
 
 // Start implements node.Lifecycle, starting up the monitoring and reporting daemon.
 func (s *Service) Start() error {
-	go s.loop()
+	// Subscribe to chain events to execute updates on
+	chainHeadCh := make(chan core.ChainHeadEvent, chainHeadChanSize)
+	headSub := s.backend.SubscribeChainHeadEvent(chainHeadCh)
+	s.headSub = headSub
```
Comment (holiman) [inline ethstats/ethstats.go:175]:
```suggestion
	s.headSub = s.backend.SubscribeChainHeadEvent(chainHeadCh)
```

File: ethstats/ethstats.go [hunk 4]
```diff
@@ -167,38 +169,30 @@ func New(node *node.Node, backend backend, engine consensus.Engine, url string)
...
+	txEventCh := make(chan core.NewTxsEvent, txChanSize)
+	txSub := s.backend.SubscribeNewTxsEvent(txEventCh)
+	s.txSub = txSub
```
Comment (holiman) [inline ethstats/ethstats.go:178]:
```suggestion
	txEventCh := make(chan core.NewTxsEvent, txChanSize)
	s.txSub = s.backend.SubscribeNewTxsEvent(txEventCh)
```

**Pull request merged**